File: ch1/Ex_1.3.4.1/task08.cpp
#include <bits/stdc++.h>                         // C++ code for task 8
using namespace std;
#define LSOne(S) ((S) & -(S))                    // notice the brackets
int main() {
  int N = 4;                                     // change 4 to 20 to match the actual problem (but it will take some time to print)
  for (int i = 0; i < (1<<N); ++i) {
    int pos = i;
    while (pos) {
      int ls = LSOne(pos);
      printf("%d ", __builtin_ctz(ls)+1);        // this idx is part of set
      pos -= ls;
    }
    printf("\n");
  }
}

File: ch1/Ex_1.3.4.1/task08.ml
open Printf

let ls_one s = s land (-s)

let rec ctz x =
	let rec helper num ans = 
		if num land 1 = 1 then ans
	 	else helper (num lsr 1) (ans+1)
	 in 
	 helper x 0

let () =
	let n = 4 in (* change 4 to 20 to match the actual problem (but it will take some time to print) *)
	for i = 0 to (1 lsl n) - 1 do 
		let pos = ref i in 
		while !pos > 0 do
			let ls = ls_one !pos in 
			pos := !pos - ls;
			printf "%d " (ctz ls)
		done;
		printf "\n"
	done